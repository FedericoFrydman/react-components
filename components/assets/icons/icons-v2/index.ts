import { V2_APPS_ICONS } from './apps';
import { V2_ARROW_ICONS } from './arrows';
import { V2_BADGES_ICONS } from './badges';
import { V2_CALENDAR_ICONS } from './calendar';
import { V2_COMMUNICATION_ICONS } from './communication';
import { V2_EDIT_ICONS } from './edit';
import { V2_FEEDBACK_ICONS } from './feedback';
import { V2_FILE_ICONS } from './file';
import { V2_FILE_TYPES_ICONS } from './file-types';
import { V2_INTERFACE_ICONS } from './interface';
import { V2_LIVE_ICONS } from './live';
import { V2_LOGOS_ICONS } from './logos';
import { V2_MARKS_ICONS } from './marks';
import { V2_MEDIA_ICONS } from './media';
import { V2_MEDIALOGO_ICONS } from './media-logos';
import { V2_MENU_ICONS } from './menu';
import { V2_NAVIGATION_ICONS } from './navigation';
import { V2_PARTNERSHIPS_ICONS } from './partnerships';
import { V2_PAYMENTS_ICONS } from './payments';
import { V2_SOCIAL_ICONS } from './social';
import { V2_SYSTEM_ICONS } from './system';
import { V2_USER_ICONS } from './user';
import { V2_WARNING_ICONS } from './warning';

export const ICON_COLORS_V2 = [
  'default',
  'B10',
  'B100',
  'B200',
  'B300',
  'G10',
  'G100',
  'G200',
  'G300',
  'N00',
  'N50',
  'N200',
  'N300',
  'N500',
  'N600',
  'N700',
  'N800',
  'N1000',
  'P10',
  'P100',
  'P200',
  'P300',
  'R10',
  'R100',
  'R200',
  'R300',
  'Y10',
  'Y100',
  'Y200',
  'Y300',
  'BRAND50',
  'BRAND100',
  'BRAND200',
  'BRAND300',
];

export type V2IconSetName =
  | 'apps'
  | 'arrow'
  | 'calendar'
  | 'communication'
  | 'edit'
  | 'file'
  | 'interface'
  | 'live'
  | 'media'
  | 'menu'
  | 'navigation'
  | 'social'
  | 'system'
  | 'user'
  | 'warning';

export const IconsV2: { [char: string]: string } = {
  ...V2_APPS_ICONS,
  ...V2_ARROW_ICONS,
  ...V2_BADGES_ICONS,
  ...V2_CALENDAR_ICONS,
  ...V2_COMMUNICATION_ICONS,
  ...V2_EDIT_ICONS,
  ...V2_FEEDBACK_ICONS,
  ...V2_FILE_ICONS,
  ...V2_FILE_TYPES_ICONS,
  ...V2_INTERFACE_ICONS,
  ...V2_LIVE_ICONS,
  ...V2_LOGOS_ICONS,
  ...V2_MARKS_ICONS,
  ...V2_MEDIA_ICONS,
  ...V2_MEDIALOGO_ICONS,
  ...V2_MENU_ICONS,
  ...V2_NAVIGATION_ICONS,
  ...V2_PARTNERSHIPS_ICONS,
  ...V2_PAYMENTS_ICONS,
  ...V2_SOCIAL_ICONS,
  ...V2_SYSTEM_ICONS,
  ...V2_USER_ICONS,
  ...V2_WARNING_ICONS,
};

export const ICON_NAMES_V2: string[] = Object.keys(IconsV2);

export type IconNamesV2 = (typeof IconsV2)[number];

export type IconColorsV2 = (typeof ICON_COLORS_V2)[number];
