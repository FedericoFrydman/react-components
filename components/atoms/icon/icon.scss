@import '../../assets/styles/brand.scss';
@import '../../assets/styles/colors.scss';

.lsux-icon-container {
  display: flex;
  flex-wrap: wrap;
}

.lsux-icon {
  filter: $filter-default;
  transition: all 0.15s ease-in-out;

  &--no-transition {
    transition: none;
  }

  &.lsux-icon--mono {
    filter: none;
  }

  &--clickable {
    cursor: pointer;
  }

  &--xsmall {
    height: 12px;
    width: 12px;
  }

  &--small {
    height: 16px;
    width: 16px;
  }

  &--medium-small {
    height: 18px;
    width: 18px;
  }

  &--medium {
    height: 24px;
    width: 24px;
  }

  &--large {
    height: 32px;
    width: 32px;
  }

  &--xlarge {
    height: 48px;
    width: 48px;
  }

  &--N00 {
    filter: $filter-white;
  }
  &--N100 {
    filter: $filter-neutral-100;
  }
  &--N300 {
    filter: $filter-neutral-300;
  }
  &--N500 {
    filter: $filter-light-gray;
  }
  &--N700 {
    filter: $filter-gray;
  }
  &--N800 {
    filter: $filter-black;
  }
  &--N1000 {
    filter: $filter-deep-black;
  }

  &--G10 {
    filter: $filter-green-10;
  }
  &--G100 {
    filter: $filter-green-100;
  }
  &--G200 {
    filter: $filter-green;
  }
  &--G300 {
    filter: $filter-green-300;
  }

  &--P10 {
    filter: $filter-purple-10;
  }
  &--P100 {
    filter: $filter-purple-100;
  }
  &--P200 {
    filter: $filter-purple-200;
  }
  &--P300 {
    filter: $filter-purple-300;
  }

  &--R10 {
    filter: $filter-red-10;
  }
  &--R100 {
    filter: $filter-red-100;
  }
  &--R200 {
    filter: $filter-red-200;
  }
  &--R300 {
    filter: $filter-red;
  }

  &--B10 {
    filter: $filter-blue-10;
  }
  &--B100 {
    filter: $filter-blue-100;
  }
  &--B200 {
    filter: $filter-blue;
  }
  &--B300 {
    filter: $filter-blue-300;
  }

  &--Y10 {
    filter: $filter-yellow-10;
  }
  &--Y100 {
    filter: $filter-yellow-100;
  }
  &--Y200 {
    filter: $filter-yellow-200;
  }
  &--Y300 {
    filter: $filter-yellow;
  }

  &--BRAND50 {
    filter: var(--brand-color-prim50-filter);
  }

  &--BRAND100 {
    filter: var(--brand-color-prim100-filter);
  }

  &--BRAND200 {
    filter: var(--brand-color-prim200-filter);
  }

  &--BRAND300 {
    filter: var(--brand-color-prim300-filter);
  }
}

.circle {
  background: $R200;
  border-radius: 50%;

  &--xsmall {
    height: 4px;
    width: 4px;
  }

  &--small {
    height: 5px;
    width: 5px;
  }

  &--medium-small {
    height: 6px;
    width: 6px;
  }

  &--medium {
    height: 7px;
    width: 7px;
  }

  &--large {
    height: 9px;
    width: 9px;
  }

  &--xlarge {
    height: 11px;
    width: 11px;
  }
}
