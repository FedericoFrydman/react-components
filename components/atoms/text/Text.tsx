import { forwardRef, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './text.scss';
import '../../assets/styles/utilities.scss';

export type TextAs = 'p' | 'span';

export type TextSize = 'body' | 'description' | 'tiny' | 'small' | 'medium' | 'large' | 'extra-large';

export type TextWeight = 'normal' | 'semibold' | 'bold';

export type TextColor = 'N1000' | 'N800' | 'N700' | 'N500' | 'N300' | 'N100' | 'N00';

export interface TextProps extends Omit<React.HTMLProps<HTMLParagraphElement>, 'css'> {
  /**
   * Render a span or a paragraph. Default 'span'.
   */
  as?: TextAs;
  /**
   * Text value to be rendered
   */
  text?: string;
  /**
   * Defines the text height
   */
  textHeight?: string;
  /**
   * Size of the text. Defaults to `medium`.
   */
  textSize?: TextSize;
  /**
   * Weight of the text
   */
  textWeight?: TextWeight;
  /**
   * Sets the `textWeight` to bold
   * @deprecated Use `fontWeight='bold'` instead. Will be removed in future verison.
   */
  bold?: boolean;
  /**
   * Sets the `textWeight` to semibold
   * @deprecated Use `fontWeight='semibold'` instead. Will be removed in future verison.
   */
  semiBold?: boolean;
  /**
   * Sets the text style to italic.
   */
  italic?: boolean;
  /**
   * Add a strikethrough to the text.
   */
  strike?: boolean;
  /**
   * Text value to be rendered
   */
  children?: string | number | ReactElement;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Color of the text
   * @deprecated Text defaults to N800 color. This prop will be removed in a future verison.
   */
  color?: TextColor;
  /**
   * Modifies the text as separator
   */
  separator?: boolean;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLParagraphElement>;
}

const TextComponent = (
  {
    as,
    bold,
    children,
    color,
    italic,
    semiBold,
    separator,
    strike,
    text,
    textHeight,
    textSize = 'medium',
    textWeight = 'normal',
    classNames = [],
    ...props
  }: TextProps,
  ref: TextProps['ref'],
) => {
  // Warn if bold or semibold props are used
  if (bold || semiBold) {
    console.warn('bold & semibold are deprecated. Please use `textWeight` instead.');
  }

  // Warn if color prop is used
  if (color) {
    console.warn('color is deprecated. Text is defaulted to $N800');
  }

  // Warn if body or description textSize values are used
  if (textSize === 'body' || textSize === 'description') {
    console.warn(
      'textSizes `body` and `description` have been deprecated. Please use `tiny`, `small`, `medium`, `large`, or `extra-large` instead.',
    );
  }

  // Set textWeight
  if (semiBold) {
    textWeight = 'semibold';
  }
  if (bold) {
    textWeight = 'bold';
  }

  const classText = 'lsux-text';
  const setHeight = { height: textHeight, overflow: 'auto' };
  const useHeight = textHeight ? setHeight : {};

  const cnText = cn(
    {
      [`${classText}`]: true,
      [`${classText}--separator`]: separator,
      [`${classText}--italic`]: italic,
      [`${classText}--strike`]: strike,
      [`${classText}--${textSize}`]: textSize,
      [`${classText}--${textWeight}`]: textWeight,
    },
    classNames,
  );

  if (as === 'span') {
    return (
      <span {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </span>
    );
  } else {
    return (
      <p {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </p>
    );
  }
};

export const Text = forwardRef<HTMLParagraphElement, TextProps>(TextComponent);
