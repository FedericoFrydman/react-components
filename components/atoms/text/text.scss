@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/colors.scss';

.lsux-text {
  margin: 0;

  // Text Weights
  &--normal {
    font-weight: normal;
  }
  &--bold {
    font-weight: bold;
  }
  &--semibold {
    font-weight: 600;
  }

  // Text Sizes
  &--tiny {
    font-size: 10px;
    font-style: normal;
    letter-spacing: -0.01em;
    line-height: 12px;
  }
  &--description,
  &--small {
    font-size: 12px;
    font-style: normal;
    letter-spacing: -0.006em;
    line-height: 16px;
  }
  &--body,
  &--medium {
    font-size: 14px;
    font-style: normal;
    letter-spacing: -0.006em;
    line-height: 18px;
  }
  &--large {
    font-size: 16px;
    font-style: normal;
    letter-spacing: -0.011em;
    line-height: 22px;
  }
  &--extra-large {
    font-size: 18px;
    font-style: normal;
    letter-spacing: -0.014em;
    line-height: 24px;
  }

  // Text Styles
  &--italic {
    font-style: italic;
  }
  &--strike {
    text-decoration: line-through;
  }

  &--no-margin {
    margin: 0;
  }

  &--separator {
    align-items: center;
    display: flex;
    text-transform: lowercase;

    &::before,
    &::after {
      background-color: $N300;
      content: '';
      flex-grow: 1;
      height: 1px;
    }

    &::before {
      margin-right: 16px;
    }

    &::after {
      margin-left: 16px;
    }
  }
}
