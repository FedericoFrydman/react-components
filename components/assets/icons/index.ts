export * from './icons-v1';
export * from './icons-v2';

import { IconsV1, ICON_COLORS_V1, ICON_NAMES_V1 } from './icons-v1';
import { IconsV2, ICON_COLORS_V2, ICON_NAMES_V2 } from './icons-v2';

export type IconSet = { [name: string]: string };

export const getLibraryVerisonByIcon = (name: string): string | null => {
  if (Object.keys(IconsV2).includes(name)) {
    return '2';
  } else if (Object.keys(IconsV1).includes(name)) {
    return '1';
  } else {
    return null;
  }
};

export const ICON_NAMES = [...ICON_NAMES_V1, ...ICON_NAMES_V2];

export const ICON_COLORS_ALL = [...ICON_COLORS_V1, ...ICON_COLORS_V2];

export type IconColors = (typeof ICON_COLORS_ALL)[number];

export const Icons = { ...IconsV1, ...IconsV2 };
