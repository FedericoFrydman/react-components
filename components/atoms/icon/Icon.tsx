import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './icon.scss';
import { Container } from '../container/Container';
import { Icons } from '../../assets/icons';
import { IconColorsV1, IconNamesV1 } from '../../assets/icons/icons-v1';
import { IconColorsV2, IconNamesV2 } from '../../assets/icons/icons-v2';

export type IconLibraryVersion = '1' | '2';

export type IconNames = IconNamesV1 | IconNamesV2;

export type IconAndBadgeSizes = 'xsmall' | 'small' | 'medium-small' | 'medium' | 'large' | 'xlarge';

export const excludedCategories: string[] = ['apps', 'badges', 'logo', 'logos', 'feedback', 'payments', 'medialogo'];

export interface IconProps
  extends Omit<React.HTMLProps<HTMLImageElement>, 'css' | 'alt' | 'src' | 'size' | 'crossOrigin'> {
  /**
   * Applies a color filter to the icon.
   */
  color?: IconColorsV1 | IconColorsV2;
  /**
   * Name of the icon to be shown.
   */
  name: IconNamesV1 | IconNamesV2;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Defines icon size and badge size.
   */
  size?: IconAndBadgeSizes;
  /**
   * Disables CSS transition on the icon.
   */
  noTransition?: boolean;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  onClick?: (e: React.MouseEvent) => void | unknown;
  /**
   * Alt tag for the HTML element
   * @default `name` paramater
   */
  alt?: string;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLImageElement>;
  /**
   * Define if icon has a Badge.
   */
  useBadge?: boolean;
}

const IconComponent = (
  {
    classNames = [],
    color = 'default',
    name,
    alt = name + '.',
    onClick,
    size = 'medium',
    noTransition,
    useBadge = false,
    ...props
  }: IconProps,
  ref: IconProps['ref'],
) => {
  const classIcon = 'lsux-icon';
  const classIconContainer = 'lsux-icon-container';

  const applyMonoClass = () => {
    let isMono = false;
    excludedCategories.forEach((c) => {
      if (name?.includes(`${c}_`)) isMono = true;
    });
    return isMono ? `${classIcon}--mono` : '';
  };

  const cnIconContainer = [`${classIconContainer}`, `flex-row`];

  const cnIcon = cn(
    {
      [`${classIcon}`]: true,
      [`${classIcon}--${color}`]: color !== 'default',
      [`${classIcon}--${size}`]: size,
      [`${classIcon}--no-transition`]: noTransition,
      [`${classIcon}--clickable`]: !!onClick,
      [applyMonoClass()]: true,
    },
    classNames,
  );

  const cnBadge = cn({
    [`circle`]: true,
    [`circle--${size}`]: size,
  });

  if (name === null || name === undefined) {
    return <div onClick={onClick} className={cnIcon}></div>;
  } else {
    return (
      <Container classNames={cnIconContainer} background="none" flexbox flexDirection="row">
        <img {...props} src={Icons[name] as string} alt={alt} className={cnIcon} ref={ref} onClick={onClick}></img>
        {useBadge && <div className={cnBadge} />}
      </Container>
    );
  }
};

export const Icon = forwardRef<HTMLImageElement, IconProps>(IconComponent);
