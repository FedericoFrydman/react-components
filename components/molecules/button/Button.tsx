import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './button.scss';
import { Icon, IconAndBadgeSizes, IconNames } from '../../atoms/icon/Icon';
import { IconColors } from '../../assets/icons';
import { Text, TextSize } from '../../atoms/text/Text';

export type ButtonShape = 'round' | 'rectangular';

export type ButtonType = 'button' | 'submit' | 'reset';

export type ButtonVariant = 'primary' | 'secondary' | 'tertiary' | 'standard' | 'destructive' | 'link';

export interface ButtonProps extends Omit<React.HTMLProps<HTMLButtonElement>, 'type' | 'css'> {
  /**
   * The text to be shown in the button
   */
  label?: string;
  /**
   * Style of the button. Values are `primary`, `secondary`, `tertiary`, `standard`, `destructive`, and `link`.
   * The following have been deprecated: `standard`, `destructive`, `link`. Please use other variants instead and set the destructive prop to true for destructive variants.
   * @default primary
   */
  variant?: ButtonVariant;
  /**
   * Indicates the button has a destructive action and sets the color to red.
   * @default false
   */
  destructive?: boolean;
  /**
   * Visually and functionally disable the button.
   * @default false
   */
  disabled?: boolean;
  /**
   * Shape of the button - either round or rectangular.
   * @default rectangular
   */
  shape?: ButtonShape;
  /**
   * Stretch the button across the full width of its parent.
   * @default false
   */
  stretch?: boolean;
  /**
   * Display an icon in addition to the text to help to identify the action. The color will be same of the text
   */
  iconLeft?: IconNames;
  /**
   *  Display an icon on the right
   */
  iconRight?: IconNames;
  /**
   * Icon size
   * @default small
   */
  iconSize?: IconAndBadgeSizes;
  /**
   * Applies a color to the icon.
   */
  iconColor?: IconColors;
  /**
   * Text size of the button
   * @default large
   */
  textSize?: TextSize;
  /**
   * The HTML button type
   * @default button
   */
  type?: ButtonType;
  /**
   * Custom class for component.
   */
  classNames?: cn.Argument;
  /**
   * Define the button width
   */
  widthLong?: string;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLButtonElement>;
  /**
   * An event id that will be used by analytics and test automation.
   */
  pplsiEventId?: string;
}

const ButtonComponent = (
  {
    variant = 'primary',
    destructive = false,
    disabled = false,
    stretch = false,
    shape = 'rectangular',
    type = 'button',
    iconLeft,
    iconRight,
    iconColor,
    iconSize = 'small',
    textSize = 'large',
    label,
    onClick,
    classNames = [],
    widthLong,
    pplsiEventId,
    ...props
  }: ButtonProps,
  ref: ButtonProps['ref'],
) => {
  // Backwards compatibility for secondary buttons
  if (variant === 'standard') {
    variant = 'secondary';
    console.warn('standard variant has been deprecated. Please use `secondary` instead.');
  }

  // Backwards compatibility for link buttons
  if (variant === 'link') {
    variant = 'tertiary';
    console.warn('link variant has been deprecated. Please use `tertiary` instead.');
  }

  // Backwards compatibility for destructive buttons
  if (variant === 'destructive') {
    destructive = true;
    variant = 'primary';
    console.warn('destructive variant has been deprecated. Please use `destructive` instead.');
  }

  const classButton = 'lsux-button';
  const cnButton = cn(
    {
      [`${classButton}`]: true,
      [`${classButton}--${variant}`]: variant,
      [`${classButton}--${shape}`]: shape,
      [`${classButton}--destructive`]: destructive,
      [`${classButton}--stretch`]: stretch,
      [`${classButton}--icon-only`]: ((!iconLeft && iconRight) || (iconLeft && !iconRight)) && !label,
    },
    classNames,
  );

  const variantColor = variant === 'primary' || destructive ? 'N00' : 'default';
  const color = iconColor ? iconColor : variantColor;
  const dynamicWidth = { width: widthLong };
  const setStyle = widthLong ? dynamicWidth : {};

  const cnText = iconLeft || iconRight ? [cn({ 'pl-3': iconLeft, 'pr-3': iconRight })] : [];

  return (
    <button
      {...props}
      type={type}
      style={{ ...setStyle, ...props.style }}
      data-pplsi-event-id={pplsiEventId}
      className={cnButton}
      disabled={disabled}
      ref={ref}
      onClick={onClick}
    >
      {iconLeft && <Icon name={iconLeft} color={color} size={iconSize} classNames={['lsux-button--icon-left']} />}
      {label && <Text textSize={textSize} textWeight="semibold" text={label} classNames={cnText} />}
      {iconRight && <Icon name={iconRight} color={color} size={iconSize} classNames={['lsux-button--icon-right']} />}
    </button>
  );
};

/**
 * The Button component enables the user to perform an action or navigate
 * to a different screen.
 */

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(ButtonComponent);
