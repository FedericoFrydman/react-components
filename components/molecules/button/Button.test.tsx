import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Button, ButtonProps, ButtonType, ButtonVariant } from './Button';
import { render, screen } from '@testing-library/react';
import { IconColors } from '../../assets/icons';
import { ICON_COLORS_V2 } from '../../assets/icons/icons-v2';
import { TextSize } from '../../atoms/text/Text';
import { IconNames } from '../../atoms/icon/Icon';

describe('Button Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;
  let pplsiEventId: string;

  beforeEach(() => {
    wrapper = shallow(<Button label="button" />);
    testId = 'test button';
    pplsiEventId = 'test prop';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the button the provided classNames', () => {
          const button = renderButton({ classNames });

          expect(button).toHaveClass(cn(classNames), { exact: false });
        });
      });

      describe('when provided variant prop', () => {
        const classNames = 'lsux-button--primary';

        it('gives the button the provided classNames', () => {
          const button = renderButton({ variant: 'primary' });

          expect(button).toHaveClass(classNames, { exact: false });
        });
      });

      describe('when is iconOnly button', () => {
        const classNames = 'lsux-button--icon-only';

        it('gives the button the provided classNames', () => {
          const button = renderButton({ iconRight: 'object_truck' });

          expect(button).toHaveClass(classNames, { exact: false });
        });
      });
    });

    describe('disabled', () => {
      let disabled: boolean;

      it('defaults to enabled', () => {
        expect(renderButton()).toBeEnabled();
      });

      describe('when true', () => {
        beforeEach(() => (disabled = true));

        it('disables the button', () => {
          expect(renderButton({ disabled })).toBeDisabled();
        });
      });

      describe('when false', () => {
        beforeEach(() => (disabled = false));

        it('the button is enabled', () => {
          expect(renderButton({ disabled })).toBeEnabled();
        });
      });
    });

    describe('iconLeft', () => {
      const iconLeftClass = `lsux-button--icon-left`;
      const iconLeftColor = `lsux-icon--R300`;
      let iconLeft: IconNames;

      describe('when "iconLeft" is present', () => {
        beforeEach(() => (iconLeft = 'edit_add_plus'));

        it(`gives the icon the ${iconLeftClass} class`, () => {
          renderButton({ iconLeft });

          const icon = screen.getByRole('img');
          expect(icon).toHaveClass(iconLeftClass);
        });

        it('passes the icon name to the icon', () => {
          renderButton({ iconLeft });

          const icon = screen.getByRole('img');
          expect(icon).toHaveAttribute('alt', iconLeft + '.');
        });
      });

      describe('when "iconLeft" is not present', () => {
        it('the button does not have the icon', () => {
          renderButton();

          const icon = screen.queryByRole('img');
          expect(icon).toBeNull();
        });
      });

      describe('when "iconLeft" has color', () => {
        it('the button does have a colored icon', () => {
          const iconColor = 'R300';
          renderButton({ iconColor, iconLeft });
          const icon = screen.getByRole('img');
          expect(icon).toHaveClass(iconLeftColor);
        });
      });
    });

    describe('iconOnly', () => {
      const iconOnlyClass = `lsux-button--icon-only`;
      let label: string;
      let iconLeft: IconNames, iconRight: IconNames;

      describe('when there is no "label"', () => {
        describe('when there is no "iconLeft" or "iconRight"', () => {
          it('does not give the button the "iconOnly" class', () => {
            expect(renderButton()).not.toHaveClass(iconOnlyClass);
          });
        });

        describe('when there is an "iconLeft"', () => {
          beforeEach(() => (iconLeft = 'object_truck'));

          it(`gives the button the ${iconOnlyClass} class`, () => {
            expect(renderButton({ iconLeft })).toHaveClass(iconOnlyClass, { exact: false });
          });
        });

        describe('when there is an "iconRight"', () => {
          beforeEach(() => (iconRight = 'object_truck'));

          it(`gives the button the ${iconOnlyClass} class`, () => {
            expect(renderButton({ iconRight })).toHaveClass(iconOnlyClass, { exact: false });
          });
        });
      });

      describe('when there is a label and a single icon', () => {
        beforeEach(() => {
          label = "Hi I'm testing";
          iconLeft = 'action_close';
        });

        it('does not give the button the icon-only class', () => {
          const button = renderButton({ iconLeft, label });

          expect(button).not.toHaveClass(iconOnlyClass);
        });
      });
    });

    describe('iconRight', () => {
      const iconRightClass = `lsux-button--icon-right`;
      let iconRight: IconNames;

      describe('when "iconRight" is present', () => {
        beforeEach(() => (iconRight = 'edit_add_plus'));

        it(`gives the icon the ${iconRightClass} class`, () => {
          renderButton({ iconRight });

          const icon = screen.getByRole('img');
          expect(icon).toHaveClass(iconRightClass);
        });

        it('passes the icon name to the icon', () => {
          renderButton({ iconRight });

          const icon = screen.getByRole('img');
          expect(icon).toHaveAttribute('alt', iconRight + '.');
        });
      });

      describe('when "iconRight" is not present', () => {
        it('the button does not have the icon', () => {
          renderButton();

          const icon = screen.queryByRole('img');
          expect(icon).toBeNull();
        });
      });
    });

    describe('label', () => {
      const label = 'my button label';

      it('is displayed when present', () => {
        renderButton({ label });

        expect(screen.getByText(label)).toBeVisible();
      });
    });

    describe('stretch', () => {
      const stretchClass = `lsux-button--stretch`;
      let stretch: boolean;

      describe('when stretch is true', () => {
        beforeEach(() => (stretch = true));

        it(`has the ${stretchClass} class`, () => {
          expect(renderButton({ stretch })).toHaveClass(stretchClass, { exact: false });
        });
      });

      describe('when stretch is false', () => {
        beforeEach(() => (stretch = false));

        it(`does not have the ${stretchClass} class`, () => {
          expect(renderButton({ stretch })).not.toHaveClass(stretchClass, { exact: false });
        });
      });
    });

    describe('textSize', () => {
      const textSizes: TextSize[] = ['body', 'description', 'tiny', 'small', 'medium', 'large', 'extra-large'];
      const label = 'My test button';
      let textSizeClass: string;

      describe('when there is a label', () => {
        it('defaults "textSize" to the "large"', () => {
          renderButton({ label });

          const labelEl = screen.getByText(label);
          expect(labelEl).toHaveClass('lsux-text--large');
        });

        function itAssignsTheRightClassForTextSize(textSize: TextSize) {
          describe(`when "textSize" is "${textSize}"`, () => {
            beforeEach(() => {
              textSizeClass = `lsux-text--${textSize}`;
            });

            it(`assigns the label the ${textSizeClass} class`, () => {
              renderButton({ label, textSize });

              const labelEl = screen.getByText(label);
              expect(labelEl).toHaveClass(textSizeClass);
            });
          });
        }

        textSizes.forEach((textSize) => itAssignsTheRightClassForTextSize(textSize));
      });
    });

    describe('type', () => {
      describe('when provided a valid button type', () => {
        const buttonTypes: ButtonType[] = ['button', 'reset', 'submit'];

        function itAssignsTheRightButtonType(type: ButtonType) {
          describe(`when the type is ${type}`, () => {
            it(`sets the type html attribute as ${type}`, () => {
              const button = renderButton({ type });

              expect(button).toHaveAttribute('type', type);
            });
          });
        }

        buttonTypes.forEach((type) => itAssignsTheRightButtonType(type));
      });

      describe('when not provided a button type', () => {
        it('defaults to the "button" type', () => {
          expect(renderButton()).toHaveAttribute('type', 'button');
        });
      });
    });

    describe('destructive', () => {
      describe('when destructive is true', () => {
        it('has the correct classname', () => {
          const button = renderButton({ destructive: true });

          expect(button).toHaveClass(`lsux-button--destructive`);
        });
      });
    });

    describe('variant', () => {
      const variants: ButtonVariant[] = ['primary', 'secondary', 'tertiary'];

      it('defaults to the primary variant', () => {
        const button = renderButton();

        expect(button).toHaveClass(`lsux-button--primary`);
      });

      // Backwards compatibility
      it(`defaults to 'secondary' when value is 'standard'`, () => {
        const button = renderButton({ variant: 'standard' });

        expect(button).toHaveClass(`lsux-button--secondary`, { exact: false });
      });

      // Backwards compatibility
      it(`defaults to 'tertiary' when value is 'link'`, () => {
        const button = renderButton({ variant: 'link' });

        expect(button).toHaveClass(`lsux-button--tertiary`, { exact: false });
      });

      // Backwards compatibility
      it(`defaults to 'primary' when value is 'destructive'`, () => {
        const button = renderButton({ variant: 'destructive' });

        expect(button).toHaveClass(`lsux-button--primary`, { exact: false });
      });

      // Backwards compatibility
      it(`sets 'destructive' to true when variant is 'destructive'`, () => {
        const button = renderButton({ variant: 'destructive' });

        expect(button).toHaveClass(`lsux-button--destructive`, { exact: false });
      });

      describe('when the button contains an icon', () => {
        const iconRight: IconNames = 'edit_add_plus';

        describe('variants with a dark background', () => {
          const darkVariants: ButtonVariant[] = ['primary'];

          function useAwhiteIconColor(variant: ButtonVariant) {
            const whiteIconClass = `lsux-icon--N00`;

            it('use a white icon color', () => {
              renderButton({ iconRight, variant });

              const img = screen.getByRole('img');
              expect(img).toHaveClass(whiteIconClass);
            });
          }

          darkVariants.forEach((variant) => useAwhiteIconColor(variant));
        });

        describe('variants with a light background', () => {
          const lightVariants: ButtonVariant[] = ['secondary', 'tertiary'];

          function usePurpleIconColor(variant: ButtonVariant) {
            const purpleIconClass = `lsux-icon-v2--BRAND200`;

            xit('use a purple icon color', () => {
              renderButton({ iconRight, variant });

              const img = screen.getByRole('img');
              const iconColorClasses = ICON_COLORS_V2.map((color) => `lsux-icon--${color}`);
              const containsIconColorClass = Object.values(img.classList).some((klass) =>
                iconColorClasses.includes(klass as IconColors),
              );
              expect(containsIconColorClass).toEqual(false);
            });
          }

          lightVariants.forEach((variant) => usePurpleIconColor(variant));
        });
      });

      function itAssignsTheRightButtonTypeClassForVariant(variant: ButtonVariant) {
        const buttonTypeClass = `lsux-button--${variant}`;

        describe(`when the variant is ${variant}`, () => {
          it(`assigns the ${buttonTypeClass} to the button`, () => {
            const button = renderButton({ variant });

            expect(button).toHaveClass(buttonTypeClass);
          });
        });
      }

      variants.forEach((variant) => itAssignsTheRightButtonTypeClassForVariant(variant));
    });

    describe('widthLong', () => {
      describe('when "widthLong" is provided', () => {
        const widthLong = `200px`;

        it('sets the button width to the provided value', () => {
          const button = renderButton({ widthLong });

          expect(button).toHaveStyle({ width: widthLong });
        });
      });
    });
  });

  describe('onClick', () => {
    const onClickSpy = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    it('when "onClick" is provided it is called when clicked', () => {
      render(<Button onClick={onClickSpy} data-testid={testId} />);
      const button = screen.getByTestId(testId);

      button.click();

      expect(onClickSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLButtonElement>();

    it('provides a ref to the button', () => {
      expect(renderButton({ ref })).toEqual(ref.current);
    });
  });

  describe('unspecified html attributes', () => {
    const altText = 'My accessible button';

    it('allows html attributes to be passed to the underlying button element', () => {
      render(<Button alt={altText} data-testid={testId} />);

      const button = screen.getByTestId(testId);

      expect(button).toHaveAttribute('alt', altText);
    });

    it('allows pplsi event id data attribute to be passed to the underlying button element', () => {
      render(<Button pplsiEventId={pplsiEventId} data-testid={testId} />);

      const button = screen.getByTestId(testId);

      expect(button).toHaveAttribute('data-pplsi-event-id', pplsiEventId);
    });
  });

  const renderButton = (props: ButtonProps = {}): HTMLElement => {
    render(<Button {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
