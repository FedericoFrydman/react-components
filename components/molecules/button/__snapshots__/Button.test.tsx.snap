// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render the component 1`] = `
<button
  className="lsux-button lsux-button--primary lsux-button--rectangular"
  disabled={false}
  style={Object {}}
  type="button"
>
  <ForwardRef(TextComponent)
    classNames={Array []}
    text="button"
    textSize="large"
    textWeight="semibold"
  />
</button>
`;
